File: 907-koko-eating-bananas/solution.py
class Solution:
    def minEatingSpeed(self, piles: List[int], h: int) -> int:
        
        def f(piles, speed):
            hours = 0
            for x in piles:
                hours += x // speed
                if x % speed > 0:
                    hours += 1
            return hours
        
        left, right = 1, 10**9

        # left bound binary search
        while (left <= right):
            mid = left + (right - left) // 2
            if (f(piles, mid) > h):
                left = mid + 1
            elif (f(piles, mid) < h):
                right = mid - 1
            elif (f(piles, mid) == h):
                right = mid - 1
        return left

File: 907-koko-eating-bananas/solution.cpp
class Solution {
public:
    int minEatingSpeed(vector<int>& piles, int h) {
        int left = 1;
        int right = 1000000000 + 1;

        while (left < right) {
            int mid = left + (right - left) / 2;
            if (f(piles, mid) <= h) {
                right = mid;
            } else {
                left = mid + 1;
            }
        }
        return left;
    }

private:
    int f(vector<int>& piles, int candidateK) {
        int hours = 0;
        for (int i = 0; i < piles.size(); i++) {
            hours += piles[i] / candidateK + (int)(piles[i] % candidateK > 0);
        }
        return hours;
    }
};
